doc/concept/build_system.md

```diff
@@ -0,0 +1,73 @@
+_Note: This document is designed to render nicely with GitHub-flavored Markdown.  My apologies in advance to everyone else._
+
+Build System Concepts
+=====================
+
+Goals of this document
+----------------------
+
+Be deleberate about choosing a build system; explain criteria, options, choices, and decisions.
+
+Strategy to choose a build system is:
+1. Enumerate good experiences to retain
+2. Enumerate bad experiences to avoid
+3. Be judiciously terse; the longer this is, the fewer people will read it
+4. Enumerate & experiment with options
+5. Make decision
+
+
+Definitions
+-----------
+
+**Contributor:** Who is a contributor?  Is there some base-level knowledge we can expect?
+
+
+Requirements
+------------
+
+* Minimal dependencies as practical, to facilitate easy building by others
+* Cross-platform
+* Any external dependencies must be freely installable (i.e., no registration, no license codes, no money; see also `CONTRIBUTING.md`)
+* Any internally satisfied dependencies must be freely runable (i.e., no license codes; see also `CONTRIBUTING.md`)
+* Is maintainable by contributors given a reasonable learning curve
+
+
+Good experiences to retain
+--------------------------
+
+_Disclaimer: A good or bad experience in any particular build system is not an implicit vote for or against that build system; these days, how you use build systems is just as important as which one you choose._
+
+**Single build system for everything:** (ABK) Used Waf to create a single build system for the entire project; fetching dependencies, configuring, building, testing, packaging, all from just `waf configure --platform=<...> && waf`.
+
+**Fully internal dependencies:** (ABK) Built an ad-hoc package manager (thin wrapper on top of `git clone`) into the project's configure stage so that the project can compile on any computer for any platform (within reason).  This includes required 3rd party libraries, required 3rd party tools, you name it.  Setting up a developer machine no longer requires any special work past SSH keys, a working compiler and such.
+
+
+Bad experiences to avoid
```

Comment:
I really like this section -- but I think there's a few more I could add. At a high level:

- Installing to a system location is hard.
- Embedding project in another project is also hard. Given for example a user would like to add the library to an existing project, doing this currently is super-hard.

Reply:
What characteristics do you see as being conducive to embedding inside another project?

In my organization, we have cpp-netlib pre-compiled by hand, and then committed in our package manager.  The downside is that when we upgrade boost, we also need to recompile cpp-netlib.  We don't upgrade boost often, so it's not a big deal.

What sort of interface do you imagine between cpp-netlib and other projects?  Compile first and interface the compiled results with a dependent project, or interface cpp-netlib's source with a dependent project, and both are built at the same time?